aspects/schema.go

```diff
@@ -138,7 +138,42 @@ func (s *StorageSchema) Validate(raw []byte) error {
 	return s.topLevel.Validate(raw)
 }
 
+// parseAlternatives tries to parse raw as a JSON list of alternatives types.
+// If the input is not a JSON list, a *json.UnmarshalTypeError is returned.
+func (s *StorageSchema) parseAlternatives(raw json.RawMessage) (*alternativesSchema, error) {
+	var alternatives []json.RawMessage
+	if err := json.Unmarshal(raw, &alternatives); err != nil {
+		return nil, err
+	}
+
+	var alt alternativesSchema
+	for _, altRaw := range alternatives {
+		schema, err := s.parse(altRaw)
+		if err != nil {
+			return nil, err
+		}
+
+		alt.schemas = append(alt.schemas, schema)
+	}
+
+	if len(alt.schemas) == 0 {
+		return nil, fmt.Errorf(`alternative type list cannot be empty`)
+	}
+
+	return &alt, nil
+}
+
 func (s *StorageSchema) parse(raw json.RawMessage) (parser, error) {
+	alt, err := s.parseAlternatives(raw)
+	if err != nil {
+		var typeErr *json.UnmarshalTypeError
+		if !errors.As(err, &typeErr) {
+			return nil, fmt.Errorf(`cannot parse aspect schema: %w`, err)
+		}
+	} else {
+		return alt, nil
+	}
```

Comment:
I find this section a little confusing. Maybe something like this?
```go
if err == nil {
	return alt, nil
}
                                                                  
var typeErr *json.UnmarshalTypeError
if !errors.As(err, &typeErr) {
	return nil, fmt.Errorf(`cannot parse aspect schema: %w`, err)
}
```

I think a comment explaining why you fallthrough and don't bubble up the error if it is a `json.UnmarshalTypeError` would be helpful.

Reply:
I see now, `raw` should be one of three types: a string, a map, or a slice.

How would you feel about a function like this:
```go
func checkJSONTypes(raw json.RawMessage) (any, error) {
	var slice []json.RawMessage
	if err := json.Unmarshal(raw, &slice); err == nil {
		return slice, nil
	}

	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err == nil {
		return m, nil
	}

	var s string
	if err := json.Unmarshal(raw, &s); err == nil {
		return s, nil
	}

	return nil, fmt.Errorf("cannot parse aspect schema")
}
```

Then maybe something like this?
```go
jsonType, err := checkJSONTypes(raw)
if err != nil {
	return nil, err
}
                                       
switch jsonValue := jsonType.(type) {
case []json.RawMessage:
	// do stuff knowing it is a list
case map[string]json.RawMessage:
	// do stuff knowing it is a map
case string:
	// do stuff knowing it is a string
}
```

Reply:
Thank you for the suggestion. I've refactored this a bit to be more explicit about the types of definition we accept and what happens